File: backend/Services/AuthService.cs
using Microsoft.EntityFrameworkCore;
using finacc.DataAccess;
using finacc.Models;
using finacc.DTOs;
using finacc.Validators;

namespace finacc.Services;


public interface IAuthService
{
    Task<Result<AuthResponse>> Login(LoginRequest request);
    Task<Result<AuthResponse>> Register(RegisterRequest request);
}


public class AuthService : IAuthService
{
    private readonly ApplicationDbContext _context;
    private readonly IPasswordHasher _passwordHasher;
    private readonly IJwtTokenService _jwtTokenService;
    private readonly AuthDataValidator _dataValidator = new();


    public AuthService(
        ApplicationDbContext context,
        IPasswordHasher passwordHasher,
        IJwtTokenService jwtTokenService)
    {
        _context = context;
        _passwordHasher = passwordHasher;
        _jwtTokenService = jwtTokenService;
    }

    public async Task<Result<AuthResponse>> Login(LoginRequest request)
    {
        // Проверяем ввод пользователя
        var loginRequestCheckResult = _dataValidator.CheckLoginRequest(request);
        if (!loginRequestCheckResult.Success)
        {
            return Result<AuthResponse>.Failure(loginRequestCheckResult.Message!);
        }

        // Ищем пользователя в БД
        var user = await _context.Users
            .FirstOrDefaultAsync(u => u.Email == request.Email);

        // Проверяем совпадения пароля
        if (user == null || !_passwordHasher.Verify(request.Password, user.PasswordHash))
        {
            return Result<AuthResponse>.Failure("Почта или пароль неверны!");
        }

        // Генерируем и создаем всю нужную информацию для ответа
        var accessToken = _jwtTokenService.GenerateAccessToken(user);
        var refreshToken = _jwtTokenService.GenerateRefreshToken();
        var userDto = new UserDto
        {
            Id = user.Id,
            Email = user.Email,
            FullName = user.FullName,
            RoleId = user.RoleId,
            OrganizationId = user.OrganizationId
        };

        var response = new AuthResponse
        {
            Token = accessToken,
            RefreshToken = refreshToken,
            UserDto = userDto
        };

        return Result<AuthResponse>.Success(response);
    }

    public async Task<Result<AuthResponse>> Register(RegisterRequest request)
    {
        var registerRequestCheckResult = _dataValidator.CheckRegisterRequest(request);
        if (!registerRequestCheckResult.Success)
        {
            return Result<AuthResponse>.Failure(registerRequestCheckResult.Message!);
        }

        // Проверяем есть ли уже пользователь с таким же email
        if (await _context.Users.AnyAsync(u => u.Email == request.Email))
        {
            return Result<AuthResponse>.Failure("Пользователь с таким email уже существует");
        }

        // Создаем нового пользователя 
        var user = new User
        {
            Email = request.Email,
            PasswordHash = _passwordHasher.Hash(request.Password),
            FullName = request.FullName,
            RoleId = null,
            OrganizationId = null,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        // Сохраняем пользователя в БД
        _context.Users.Add(user);
        await _context.SaveChangesAsync();

        // Генерируем и создаем всю нужную информацию для ответа
        var accessToken = _jwtTokenService.GenerateAccessToken(user);
        var refreshToken = _jwtTokenService.GenerateRefreshToken();
        var userDto = new UserDto
        {
            Id = user.Id,
            Email = user.Email,
            FullName = user.FullName,
            RoleId = user.RoleId,
            OrganizationId = user.OrganizationId
        };

        var response = new AuthResponse
        {
            Token = accessToken,
            RefreshToken = refreshToken,
            UserDto = userDto
        };

        return Result<AuthResponse>.Success(response);
    }
}

File: frontend/src/services/authService.js
import api from './api';

/**
 * Приводит ответ POST /auth/login и /auth/register к виду { token, user }.
 * Бэкенд: плоский AuthResponse (token, id, email, fullName, roleId, organizationId, roleName).
 * Поддерживается устаревший вариант с вложенным userDto.
 *
 * @param {object} raw - response.data от axios
 * @returns {{ token: string, user: object } | null}
 */
export function normalizeAuthSession(raw) {
  if (!raw || typeof raw !== 'object') return null;

  const token = raw.token;
  if (!token) return null;

  if (raw.id != null && raw.email != null) {
    return {
      token,
      user: {
        id: raw.id,
        email: raw.email,
        fullName: raw.fullName ?? '',
        roleId: raw.roleId ?? null,
        organizationId: raw.organizationId ?? null,
        roleName: raw.roleName ?? null,
      },
    };
  }

  const nested = raw.userDto ?? raw.UserDto;
  if (nested && nested.id != null && nested.email != null) {
    return {
      token,
      user: {
        id: nested.id,
        email: nested.email,
        fullName: nested.fullName ?? '',
        roleId: nested.roleId ?? null,
        organizationId: nested.organizationId ?? null,
        roleName: nested.roleName ?? null,
      },
    };
  }

  return null;
}

/**
 * Вход пользователя
 * @param {Object} userData - { email, password }
 * @returns {Promise<object>} - тело ответа API (плоский AuthResponse)
 */
export async function login(userData) {
  try {
    const response = await api.post('/auth/login', userData);
    return response.data;
  } catch (error) {
    const errorMessage =
      error.response?.data?.detail ||
      error.response?.data?.title ||
      error.message ||
      'Ошибка при входе';

    throw new Error(errorMessage);
  }
}

/**
 * Регистрация
 * @param {Object} userData - { email, fullName, password }
 * @returns {Promise<object>} - тело ответа API (плоский AuthResponse)
 */
export async function register(userData) {
  try {
    const response = await api.post('/auth/register', userData);
    return response.data;
  } catch (error) {
    console.error(error);
    const errorMessage =
      error.response?.data?.detail ||
      error.response?.data?.title ||
      error.message ||
      'Ошибка при регистрации';

    throw new Error(errorMessage);
  }
}
